pkg/snapshot/snapshotter/snapshotter.go

```diff
@@ -740,3 +751,57 @@ func (ssr *Snapshotter) checkSnapstoreSecretUpdate() bool {
 	snapstoreHash[ssr.snapstoreConfig.Provider] = newSnapstoreSecretHash
 	return true
 }
+
+// IsFullSnapshotRequiredAtStartup checks whether to take a full snapshot or not during the startup of backup-restore.
+func (ssr *Snapshotter) IsFullSnapshotRequiredAtStartup(timeWindow float64) bool {
+	if ssr.PrevFullSnapshot == nil || ssr.PrevFullSnapshot.IsFinal || time.Since(ssr.PrevFullSnapshot.CreatedOn).Hours() > timeWindow {
+		return true
+	}
+
+	if !ssr.WasScheduledFullSnapshotMissed(timeWindow) {
+		return false
+	}
+	return ssr.IsNextFullSnapshotBeyondTimeWindow(timeWindow)
+}
+
+// WasScheduledFullSnapshotMissed determines whether the preceding full-snapshot was missed or not.
+func (ssr *Snapshotter) WasScheduledFullSnapshotMissed(timeWindow float64) bool {
+	now := time.Now()
+	nextSnapSchedule := ssr.schedule.Next(now)
+
+	if miscellaneous.GetPrevScheduledSnapTime(nextSnapSchedule, timeWindow) == ssr.PrevFullSnapshot.CreatedOn {
+		ssr.logger.Info("previous full snapshot was taken at scheduled time, skipping the full snapshot at startup")
+		return false
+	}
+	return true
+}
+
+// IsNextFullSnapshotBeyondTimeWindow determines whether the next scheduled full snapshot will exceed the given time window or not.
+func (ssr *Snapshotter) IsNextFullSnapshotBeyondTimeWindow(timeWindow float64) bool {
+	now := time.Now()
+	nextSnapSchedule := ssr.schedule.Next(now)
+	timeLeftToTakeNextSnap := nextSnapSchedule.Sub(now)
+
+	return timeLeftToTakeNextSnap.Hours()+time.Since(ssr.PrevFullSnapshot.CreatedOn).Hours() > timeWindow
+}
+
+// GetFullSnapshotMaxTimeWindow returns the maximum time period in hours for which backup-restore must take atleast one full snapshot.
+func (ssr *Snapshotter) GetFullSnapshotMaxTimeWindow(fullSnapScheduleSpec string) float64 {
+	// Split on whitespace.
+	schedule := strings.Fields(fullSnapScheduleSpec)
+	if len(schedule) < 5 {
+		return defaultFullSnapMaxTimeWindow
+	}
+
+	if schedule[dayOfMonth] == "*" && schedule[dayOfWeek] == "*" && !strings.Contains(schedule[hour], "/") {
+		return defaultFullSnapMaxTimeWindow
+	} else if schedule[dayOfWeek] != "*" {
+		return defaultFullSnapMaxTimeWindow * 7
+	} else if schedule[dayOfMonth] == "*" && schedule[dayOfWeek] == "*" && strings.Contains(schedule[hour], "/") {
+		if timeWindow, err := strconv.ParseFloat(schedule[hour][strings.Index(schedule[hour], "/")+1:], 64); err == nil {
+			return timeWindow
+		}
+	}
+
+	return defaultFullSnapMaxTimeWindow
```

Comment:
```suggestion
	if schedule[dayOfWeek] != "*" {
		return defaultFullSnapMaxTimeWindow * 7
	}

	if schedule[dayOfMonth] == "*" && strings.Contains(schedule[hour], "/") {
		if timeWindow, err := strconv.ParseFloat(schedule[hour][strings.Index(schedule[hour], "/")+1:], 64); err == nil {
			return timeWindow
		}
	}

	return defaultFullSnapMaxTimeWindow
```

Reply:
I have addressed but in 2nd condition I still prefer to check `schedule[dayOfWeek] == "*"` as our function is not handling all permutations of cron better to have full check.
```go
if schedule[dayOfMonth] == "*" && schedule[dayOfWeek] == "*" && strings.Contains(schedule[hour], "/") {
		if timeWindow, err := strconv.ParseFloat(schedule[hour][strings.Index(schedule[hour], "/")+1:], 64); err == nil {
			return timeWindow
		}
	}
```